File: webook/pkg/redislock/fence.go
package redislock

import (
	"context"
	"fmt"
)

// acquireFencing 带 fencing 的单次获取：全新获取时在同一 Lua 内原子 INCR 单调计数器，
// 令牌随句柄返回（Fence()）；重入不 bump（沿用首次令牌）；被占返回 (nil,false,nil)。
//
// 为什么需要 fencing（唯一真安全）：持有者 GC/STW 暂停 → 锁 TTL 过期被别人拿走 →
// 暂停结束仍以为持锁去写 → 双写。锁本身挡不住这个时序，唯一解 = 单调令牌 + 资源侧校验。
func (c *RedisClient) acquireFencing(ctx context.Context, key, token string, cfg *lockConfig) (RedisLock, bool, error) {
	raw, err := fenceScript.Run(ctx, c.cmd, []string{lockKey(key), fenceKey(key)}, cfg.leaseMs(), token).Slice()
	if err != nil {
		return nil, false, err
	}
	if len(raw) != 2 {
		return nil, false, fmt.Errorf("redislock: 非法 fencing 获取结果 %v", raw)
	}
	status, _ := raw[0].(int64)
	fence, _ := raw[1].(int64)
	if status != -1 {
		return nil, false, nil // 被占，status 为剩余 pttl
	}
	return newLock(c, key, token, fence, cfg), true, nil
}

// FenceAccepted 资源侧 fencing 校验（应用层写法，§3.3 "二选一"之一）：
// 写被保护资源前，把本次 Fence() 令牌与资源已存的最大令牌比对，只有严格更大才放行；
// 更小 / 相等说明有更新的持有者已写过（自己多半是 GC 暂停后的过期持有者），必须拒绝。
//
//	incoming := lk.Fence()
//	if !redislock.FenceAccepted(incoming, stored.LastFence) { return ErrStaleFence }
//	// 放行后落库时把 incoming 持久化为新的 LastFence
//
// 另一种等价写法（DB 条件写，把校验合并进 SQL，天然原子、免读-改-写竞态）：
//
//	UPDATE res SET data = ?, fence = ? WHERE id = ? AND fence < ?  -- 影响行数=0 → 拒绝
//
// 未接任一资源侧校验 = 没上安全锁，只是"大概率互斥"（cron 幂等重算属可接受的 best-effort）。
func FenceAccepted(incoming, stored int64) bool {
	return incoming > stored
}

File: webook/pkg/redislock/scripts/fence.lua
-- 带 fencing 的获取。返回数组 {status, fence}：
--   status: -1=成功（首次或重入）；>=0=被占（剩余 pttl 作等待提示）
--   fence : 全新获取返回新的单调令牌；重入 / 被占返回 0（重入沿用句柄缓存的首次令牌）
-- 单调计数器 KEYS[2] 只在全新获取时 INCR，且不设过期——持久保存，过期会导致单调断裂（§3.3 风险）。
-- KEYS[1]=redislock:{k}:lock  KEYS[2]=redislock:{k}:fence
-- ARGV[1]=leaseMs  ARGV[2]=ownerToken
if redis.call('exists', KEYS[1]) == 0 then
    redis.call('hincrby', KEYS[1], ARGV[2], 1)
    redis.call('pexpire', KEYS[1], ARGV[1])
    local f = redis.call('incr', KEYS[2]) -- 全新获取才 bump
    return { -1, f }
elseif redis.call('hexists', KEYS[1], ARGV[2]) == 1 then
    redis.call('hincrby', KEYS[1], ARGV[2], 1)
    redis.call('pexpire', KEYS[1], ARGV[1])
    return { -1, 0 } -- 重入不 bump
end
local ttl = redis.call('pttl', KEYS[1])
if ttl < 0 then
    ttl = tonumber(ARGV[1])
end
return { ttl, 0 }